### DISCUSSION TITLE
Are there plans to add unit testing / auto tests?

### DISCUSSION THREAD

**Discussion by davehouser1** [warmuuh/milkman#191] [General]
Having the ability to perform unit tests / auto testing with a suite of requests is a must have. Would be great if your tool included this feature. I strongly recommend including the following sub features into the testing portion if you decide to include this feature:

- Ability to create "unit tests", which is to say a specific request to run, then assert. I think there are modules you can use to allow users to use JavaScript to confirm a test ran properly. You can use frameworks like [Mocha](https://mochajs.org/) framework, and [Chai](https://www.chaijs.com/api/bdd/) for assertions.
- Ability to create test suites, which is an aggregation of unit tests.
- Ability to create test suites of suites, combining suites together.
- Support duplicating unit tests in a suite.
- Support duplicating tests suites.
- Support re-arraigning unit tests in a suite.
- Support for async requests in a given unit test. Just because a request returned a 200OK does not mean the work is complete on the backend. There has to be someway to perform more operations to check if async work is complete. There can be a variety of of ways to assert this. In most cases Iv seen there maybe another address/route that one can check job status of a job.
- Support for access to environmental variables in tests. Or create a new environmental variable set for a suite or suite of suites.
- Show how long each unit test took to complete
- Support to show total time all tests in a suite took to run.
- The number of tests can become cumbersome, add a way to search for each unit test in a suite.
- Ability to "play" or run any unit test along without having to run the whole suite.
- Hope to see this one day.

**Comment by warmuuh** [comment]
This text looks weirdly copy&pasted...are you a real human? :)

In any case: there is a Test Plugin already. There is also a Script Plugin, so you can use chai or mocha.
Besides that, No plans to add anything else.

If you fancy, any contributions are very welcome

Sidenote: a "must have" for one person is a "feature bloat" to another...

**Reply by davehouser1** [reply]
I am, and I did copy and paste from another post I made for [apidash. ](https://github.com/foss42/apidash) discussion. 
I did this as these are features I desperately need in another API REST client so I can migrate away from Insomnia. 
Total agree the "must have" / "feature bloat" comparison. My "must" was a byproduct of me trying to get away form Insomnia. 
Glad to hear there is a test plugin, I will take a look and see how to use it.
